higher_health/views.py

```diff
@@ -72,29 +86,79 @@ def get_initial(self):
-            data = {
-                "risk_level": triage.risk,
-                "first_name": triage.first_name,
-                "last_name": triage.last_name,
-                "timestamp": triage.timestamp,
-                "msisdn": triage.hashed_msisdn,
-            }
-            return render(request, "healthcheck_receipt.html", data)
-    except Covid19Triage.DoesNotExist:
-        pass
+def staff_check(user):
+    return not user.is_staff
+
 
+@login_required
+@user_passes_test(staff_check)
+def healthcheck_receipt(request):
+    triage = Covid19Triage.objects.filter(user=request.user).last()
+    if triage:
+        data = {
+            "risk_level": triage.risk,
+            "first_name": triage.first_name,
+            "last_name": triage.last_name,
+            "timestamp": triage.timestamp,
+            "msisdn": triage.hashed_msisdn,
+        }
+        return render(request, "healthcheck_receipt.html", data)
     return HttpResponseRedirect("/")
 
 
-def healthcheck_login(request):
-    form = HealthCheckLogin()
-    return render(request, "healthcheck_receipt.html", {"form": form})
+class HealthCheckLoginView(generic.FormView):
+    form_class = HealthCheckLogin
+    template_name = "login.html"
+    success_url = reverse_lazy("healthcheck_otp")
+
+    def send_otp_sms(self, user):
+        otp = "".join(secrets.choice(string.digits) for _ in range(6))
+        h = hmac.new(settings.SECRET_KEY.encode(), otp.encode(), digestmod=sha256)
+        otp_hash = base64.b64encode(h.digest()).decode()
+        self.request.session["otp_hash"] = otp_hash
+
+        # TODO: send OTP via SMS
+        # print(user.username, otp, otp_hash)
+
+    def form_valid(self, form):
+        data = form.cleaned_data
+        msisdn = data.get("msisdn")
+        self.request.session["msisdn"] = msisdn
+        user, created = User.objects.get_or_create(username=msisdn)
+
+        self.send_otp_sms(user)
+        return super().form_valid(form)
+
+
+class HealthCheckOTPView(generic.FormView):
+    form_class = HealthCheckOTP
+    template_name = "otp.html"
+    success_url = "/"
+
+    def dispatch(self, request, *args, **kwargs):
+        if "msisdn" not in request.session:
+            return HttpResponseRedirect("/")
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["request"] = self.request
+        return kwargs
+
+    def get_context_data(self, **kwargs):
+        ctx = super().get_context_data(**kwargs)
+        if self.request.method == "GET" and self.request.session.get("msisdn"):
+            msisdn = self.request.session.get("msisdn")
+            ctx["hashed_msisdn"] = msisdn[:3] + "*" * 5 + msisdn[-4:]
+        return ctx
+
+    def form_valid(self, form):
+        msisdn = self.request.session.get("msisdn")
+        user = User.objects.get(username=msisdn)
+        login(self.request, user)
```

Comment:
We should remove the OTP from the session as well, so that each OTP can only be used once.

Reply:
Now using `session.pop()`